[RISCV] custom scmp(x,0) and scmp(0,x) lowering for RVV

camel-cdr opened the pull request:
I noticed that the current codegen for scmp(x,0) and scmp(0,x), also known as sign(x) and -sign(x), isn't optimal for RVV.
It produces a four instruction sequence of four instructions
```
    vmsgt.vi + vmslt.vi + vmerge.vim + vmerge.vim
```
for SEW<=32 and three instructions for SEW=64.
```
    scmp(0,x): vmsgt.vi + vsra.vx + vor.vi
    scmp(x,0): vmsgt.vi + vsrl.vx + vmerge.vim
```

This patch introduces a new lowering for all values of SEW which expresses the above in SelectionDAG Nodes.
This maps to two arithmetic instructions and a vector register move:
```
    scmp(0,x): vmv.v.i/v + vmsgt.vi + masked vsra.vi/vx
    scmp(x,0): vmv.v.i/v + vmsgt.vi + masked vsrl.vi/vx
```
These clobber v0, need to have a different destination than the input and need to use an additional GPR for SEW=64.
For the SEW<=32 and scmp(x,0) case a slightly different lowering was chooses:
```
    scmp(x,0): vmin.vx + vsra.i + vor.vv
```
This doesn't clobber v0, but uses a single GPR.
I deemed using a single GPR slightly better than clobbering v0 (SEW<=32), but using two GPRs as worse than using one GPR and clobbering v0. But I haven't done any empirical tests.

---

I'm not sure why the fixed-vectors tests are so messed up, so I marked this as a draft for now.

This type of lowering is also advantageous for SVE and AVX512 and could be generically implemented in `TargetLowering::expandCMP`, but you need to know of integer shift instructions of the element size are available.

Here are the alive2 transforms:
* [scmp(x,0) SEW<=32 variant](https://alive2.llvm.org/ce/z/_NZgiz)
* [scmp(x,0) SEW>32 variant](https://alive2.llvm.org/ce/z/9Nhhtk)
* [scmp(0,x)](https://alive2.llvm.org/ce/z/BrxiTs)

camel-cdr committed:
[RISCV] add scmp RVV CodeGen tests

topperc commented:
Please put the tests in a separate commit within the PR so we can see the original codegen and the change.

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -8223,6 +8225,35 @@ SDValue RISCVTargetLowering::LowerOperation(SDValue Op,
+      }
+      // scmp(lhs, 0) -> vmerge.vi(vmsgt.vi(rhs,0), vsra.vx(lhs,SEW-1), 1)
+      return DAG.getSelectCC(DL, LHS, Zero, Sra, One, ISD::SETGT);
+    } else if (ISD::isConstantSplatVectorAllZeros(LHS.getNode())) {
```
No else after return.

camel-cdr commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -8223,6 +8225,35 @@ SDValue RISCVTargetLowering::LowerOperation(SDValue Op,
+    } else if (ISD::isConstantSplatVectorAllZeros(LHS.getNode())) {
+      // scmp(0, rhs) -> vmerge.vi(vmsgt.vi(rhs,0), vsrl.vi/vx(rhs,SEW-1), -1)
+      SDValue Srl = DAG.getNode(ISD::SRL, DL, VT, RHS, Shift);
+      return DAG.getSelectCC(DL, RHS, Zero, Srl, MinusOne, ISD::SETGT);
```
Ok, I tried:
```c
SDValue Setcc = DAG.getSetCC(DL, VT, LHS, Zero, ISD::SETGT);
return DAG.getSelect(DL, VT, Setcc, Sra, One);
```
but that didn't seem to work:
```
LLVM ERROR: Cannot select: t19: nxv2i64 = vselect t18, t16, t22
  t18: nxv2i64 = setcc t2, t24, setgt:ch
    t2: nxv2i64,ch = CopyFromReg t0, Register:nxv2i64 %0
    t24: nxv2i64 = RISCVISD::VMV_V_X_VL undef:nxv2i64, Constant:i64<0>, Register:i64 $x0
  t16: nxv2i64 = sra t2, t23
    t2: nxv2i64,ch = CopyFromReg t0, Register:nxv2i64 %0
    t23: nxv2i64 = RISCVISD::VMV_V_X_VL undef:nxv2i64, Constant:i64<63>, Register:i64 $x0
  t22: nxv2i64 = RISCVISD::VMV_V_X_VL undef:nxv2i64, Constant:i64<1>, Register:i64 $x0
In function: scmp_i64z64
```
I also tried using ISD::VSELECT instead of getSelect, but that didn't work either not sure what else I can do.

topperc replied:
You need to use an i1 vector type for the setcc result. You can call getSetccResultType.

camel-cdr committed:
[RISCV] custom scmp(x,0) and scmp(0,x) lowering for RVV

The current codegen for scmp(x,0) and scmp(0,x), also known as sign(x)
and -sign(x), isn't optimal for RVV.
It produces a four instruction sequence of for instructions
    vmsgt.vi + vmslt.vi + vmerge.vim + vmerge.vim
for SEW<=32 and three instructions for SEW=64.
    scmp(0,x): vmsgt.vi + vsra.vx + vor.vi
    scmp(x,0): vmsgt.vi + vsrl.vx + vmerge.vim

This patch introduces a new lowering for all values of SEW which
expresses the above in SelectionDAG Nodes.
This maps to two arithmetic instructions and a vector register move:
    scmp(0,x): vmv.v.i/v + vmsgt.vi + masked vsra.vi/vx
    scmp(x,0): vmv.v.i/v + vmsgt.vi + masked vsrl.vi/vx
These clobber v0, need to have a different destination than the input
and need to use an additional GPR for SEW=64.
For the SEW<=32 and scmp(x,0) case a slightly different
lowering was chooses:
    scmp(x,0): vmin.vx + vsra.i + vor.vv
This doesn't clobber v0, but uses a single GPR.
We deemed using a single GPR slightly better than clobbering v0
(SEW<=32), but using two GPRs as worse than using one GPR and
clobbering v0.

camel-cdr commented:
I initially had a mistake in the conversion, it should now match the alive2 code.

lukel97 reviewed (commented):
Makes sense to me.

I wish we could write this as a combine on `vselect (setcc 0, n, gt), 1, 0 -> shl n, sew - 1`, but given that the fixed and scalable nodes diverge I think custom lowering scmp is as good as we're going to get.

lukel97 commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -8223,6 +8225,40 @@ SDValue RISCVTargetLowering::LowerOperation(SDValue Op,
+    if (ISD::isConstantSplatVectorAllZeros(RHS.getNode())) {
+      SDValue Sra = DAG.getNode(ISD::SRA, DL, VT, LHS, Shift);
+      if (SEW <= 32) {
+        // scmp(lhs, 0) -> vor.vv(vsra.vi(lhs,SEW-1), vmin.vx(lhs,1))
```
Using an extra GPR to avoid clobbering V0 for SEW=64 sounds reasonable to me. If this happens in a loop the li is likely going to be hoisted out anyway.

camel-cdr replied:
OK, I made the SEW=64 case also use the vor+vsra+vmin lowering. I originally thought that using two GPRs may be worse than a v0 clobber, if it causes a GPR spill, but I didn't consider that you can use a single GPR and load-immediates in the loop if you are starved of GPRs.

lukel97 reviewed (approved):
LGTM, just some nits

camel-cdr committed:
Merge branch 'main' into rvv_scmp
